# Финальный проект по SQL

Коронавирус застал мир врасплох, изменив привычный порядок вещей.  
В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.  

Наша компания решила быть на волне и купила крупный сервис для чтения книг по подписке.  


**Цель исследования:** в базе данных информация о книгах, издательствах, авторах, а также пользовательские обзоры книг.  
Наша первая задача как аналитика — проанализировать базу данных. Эти данные помогут сформулировать ценностное предложение для нового продукта. 
 
 **Задачи:**  
1. Посчитать, сколько книг вышло после 1 января 2000 года;  
2. Для каждой книги посчитать количество обзоров и среднюю оценку;  
3. Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;  
3. Определить автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;  
4. Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.  

<a id='intro'></a>

## Описание данных

**Таблица** `books`  

Содержит данные о книгах:  

- `book_id` — идентификатор книги;  
- `author_id` — идентификатор автора;  
- `title` — название книги;  
- `num_pages` — количество страниц;  
- `publication_date` — дата публикации книги;  
- `publisher_id` — идентификатор издателя.  

**Таблица** `authors`  

Содержит данные об авторах:  
- `author_id` — идентификатор автора;  
- `author` — имя автора.  

**Таблица** `publishers`  

Содержит данные об издательствах:  
- `publisher_id` — идентификатор издательства;  
- `publisher` — название издательства;  

**Таблица** `ratings`  

Содержит данные о пользовательских оценках книг:  
- `rating_id` — идентификатор оценки;  
- `book_id` — идентификатор книги;  
- `username` — имя пользователя, оставившего оценку;  
- `rating` — оценка книги.  

**Таблица** `reviews`  

Содержит данные о пользовательских обзорах:    
- `review_id` — идентификатор обзора;  
- `book_id` — идентификатор книги;  
- `username` — имя автора обзора;  
- `text` — текст обзора.  

[Схема базы данных](https://disk.yandex.ru/i/xOI0GMFkUFJqqA)

## Загрузка библиотек

[Начало тетрадки](#intro)

In [1]:
import pandas as pd
from sqlalchemy import create_engine

## Подключение к базе данных

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Выполнение SQL запроса и осмотр данных

### Таблица `books`


**Содержит данные о книгах:**   

`book_id` — идентификатор книги;  
`author_id` — идентификатор автора;  
`title` — название книги;  
`num_pages` — количество страниц;  
`publication_date` — дата публикации книги;  
`publisher_id` — идентификатор издателя.  

In [3]:
query = '''
SELECT 
    *
FROM books
'''

In [4]:
pd.io.sql.read_sql(query, con = engine) 

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

В базе данных находится информация о 1000 ед. книг.  

[Начало тетрадки](#intro)

### Таблица `authors`

**Содержит данные об авторах:**  

`author_id` — идентификатор автора;  
`author` — имя автора.  

In [5]:
query = '''
SELECT 
    author_id,
    author
FROM authors
'''

In [6]:
pd.io.sql.read_sql(query, con = engine) 

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

В базе данных содержится информация о 636 авторах книг.

### Таблица `publishers`

**Содержит данные об издательствах:**  

`publisher_id` — идентификатор издательства;  
`publisher` — название издательства;  

In [7]:
query = '''
SELECT 
    *
FROM publishers
'''

In [8]:
pd.io.sql.read_sql(query, con = engine) 

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

В базе данных содержится информация о 340 издательствах.

### Таблица `ratings`

**Содержит данные о пользовательских оценках книг:**  

`rating_id` — идентификатор оценки;  
`book_id` — идентификатор книги;  
`username` — имя пользователя, оставившего оценку;  
`rating` — оценка книги.  

In [9]:
query = '''
SELECT 
    *
FROM ratings
'''

In [10]:
pd.io.sql.read_sql(query, con = engine) 

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

В базе данных содержится информация о 6456 оценках книг.

### Таблица `reviews`

**Содержит данные о пользовательских обзорах на книги:**  

`review_id` — идентификатор обзора;  
`book_id` — идентификатор книги;  
`username` — имя пользователя, написавшего обзор;  
`text` — текст обзора.  

In [11]:
query = '''
SELECT 
    *
FROM reviews
'''

In [12]:
pd.io.sql.read_sql(query, con = engine) 

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

В базе данных находится информация о 2793 ревью на книги.

## Выполнение заданий

### Посчитаем, сколько книг вышло после 1 января 2000 года?

In [13]:
query = '''
SELECT 
    COUNT(*) AS book_count
FROM 
    books
WHERE
    publication_date > '2000-01-01'
'''

In [14]:
pd.io.sql.read_sql(query, con = engine)

book_count
0         819

**Вывод:**  
После 1 января 2000 года вышло книг 819 ед.

### Для каждой книги посчитаем количество обзоров и среднюю оценку

Для каждой книги посчитаем количество обзоров и средний рейтинг, сортируем по количеству обзоров:

In [15]:
query = '''

SELECT
     ratings.book_id AS book_id,
     COUNT(reviews.text) AS total_reviews,
     AVG(ratings.rating) AS mean_rating
FROM
    ratings
LEFT JOIN reviews ON 
   (reviews.book_id = ratings.book_id
    AND reviews.username = ratings.username)
GROUP BY 
      ratings.book_id
ORDER BY 
    total_reviews DESC
'''
pd.io.sql.read_sql(query, con = engine)

book_id  total_reviews  mean_rating
0        948              7     3.662500
1        695              6     4.081081
2        299              6     4.287500
3        696              6     3.830508
4        734              6     4.206897
..       ...            ...          ...
995      672              0     5.000000
996      808              0     3.000000
997      221              0     4.000000
998      387              0     4.000000
999       83              0     3.666667

[1000 rows x 3 columns]

**Вывод** Так то лучше, все книги вошли в обзор. Масимальное количество обзоров получилось 7. 

Для каждой книги посчитаем количество обзоров и средний рейтинг, сортируем по рейтингу:

In [16]:
query = '''
SELECT
     ratings.book_id AS book_id,
     COUNT(reviews.text) AS total_reviews,
     AVG(ratings.rating) AS mean_rating
FROM
     ratings
LEFT JOIN reviews ON 
        (reviews.book_id = ratings.book_id
            AND reviews.username = ratings.username)
GROUP BY 
     ratings.book_id
ORDER BY 
     mean_rating DESC
'''
pd.io.sql.read_sql(query, con = engine)

book_id  total_reviews  mean_rating
0        518              2         5.00
1        732              2         5.00
2        347              2         5.00
3        610              2         5.00
4        330              1         5.00
..       ...            ...          ...
995      915              3         2.25
996      371              2         2.00
997      316              2         2.00
998      202              3         2.00
999      303              2         1.50

[1000 rows x 3 columns]

**Вывод:** Максимальный рейтинг 5.	

### Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры

In [17]:
query = '''
SELECT
     p.publisher AS publisher_name,
     COUNT(Sub_b.book_id) AS books_count
FROM
    publishers p
LEFT JOIN (SELECT *
           FROM
           books b
           WHERE
              b.num_pages > 50) AS Sub_b
ON Sub_b.publisher_id = p.publisher_id
GROUP BY
      p.publisher
ORDER BY
COUNT(Sub_b.book_id) DESC;
          '''

In [18]:
pd.io.sql.read_sql(query, con = engine)

publisher_name  books_count
0                       Penguin Books           42
1                             Vintage           31
2            Grand Central Publishing           25
3                    Penguin Classics           24
4                              Bantam           19
..                                ...          ...
335  Simon & Schuster Childrens Books            0
336  Atheneum Books for Young Readers            0
337                        Tanglewood            0
338                      Golden Books            0
339                     Golden/Disney            0

[340 rows x 2 columns]

Издательство `Penguin Books` выпустило самое большое количество книг

### Определим автора с самой высокой средней оценкой книг — учитываем только книги с 50 и более оценками

In [19]:
query = '''
SELECT
    authors.author,
    AVG(subquery_ratings.avg_rating) AS avg_rating
FROM
    books
LEFT JOIN (
           SELECT 
                book_id,
                AVG(rating) AS avg_rating,
                COUNT(rating_id) AS rating_count
           FROM
                ratings
           GROUP BY
                book_id
           ) AS subquery_ratings ON subquery_ratings.book_id = books.book_id
LEFT JOIN authors ON authors.author_id = books.author_id
WHERE
    rating_count >= 50
GROUP BY
    author
ORDER BY
    avg_rating DESC
LIMIT 1
'''

In [20]:
pd.io.sql.read_sql(query, con = engine)

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.283844

**Вывод:** Не один, а соавторы J.K Rowling/Mary GrandPré получили самый высокий средний рейтинг.

### Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [21]:
query = '''              -- выведем среднее по количеству обзоров, округление выполним функцией round
SELECT 
    round(avg(count)) AS rounded_avg_review    
FROM
    (SELECT
        COUNT(text)
    FROM 
        reviews
    FULL OUTER JOIN (SELECT username, COUNT(rating_id) as cnt FROM ratings GROUP BY username) AS count_rating
    ON count_rating.username = reviews.username where cnt > 50
    GROUP BY reviews.username) as count;                                  
'''

In [22]:
pd.io.sql.read_sql(query, con = engine)

rounded_avg_review
0                24.0

**Вывод:** Получилось среднее количество обзоров - 24 от пользователей, которые поставили больше 50 оценок.

## Чек лист готовности

- [x]  Описаны данные
- [x]  Загружены библиотека
- [x]  Подключение к базе данных
- [x]  Выполнены SQL запросы и осмотрены данные  
- [x]  Выполнены задания:
- [x]    -посчитано, сколько книг вышло после 1 января 2000 года?
- [x]    -для каждой книги посчитано количество обзоров и среднюю оценку
- [x]    -определено издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключили из анализа брошюры
- [x]    -определены авторы с самой высокой средней оценкой книг — учитывали только книги с 50 и более оценками
- [x]    -посчитано среднее количество обзоров от пользователей, которые поставили больше 50 оценок 
- [x] Сделаны выводы